src/upload.rs

```diff
@@ -70,7 +72,16 @@ impl SnapshotUploader {
                 .to_string_lossy()
                 .to_string()
         });
-        let file_size = self.file_size(path).await?;
+
+        let file_meta = fs::metadata(path)
+            .await
+            .context(error::ReadFileMetadata { path })?;
+
+        let file_size = if file_meta.file_type().is_block_device() {
+            get_block_device_size(path).context(error::GetBlockDeviceSize { path })?
+        } else {
+            self.file_size(&file_meta).await?
+        };
```

Comment:
So for upload this is the only code that differs between block devices and files, and thus we don't need a trait with implementations for upload (like we do for download)?

Reply:
Right, it's just the size calculation that differs here.